### PR TITLE
add optional custom config file

### PR DESCRIPTION
**Pull request by thehunmonkgroup**
This adds the option for a user to write a `${HOME}/.newsletter-cli.js`
config file, which, at the moment, allows customizing the subject,
styles, and body of the generated template.

My JS skills are pretty good, but I've never written a line of TypeScript before -- not totally sure I got all the syntax the best it can be, particularly the conditional include of the user config file.

Also not 100% sure about the names I used in the config file. Perhaps better namespacing like this?

```javascript
const config = {}
config.template = {}
config.template.subject = "SYSTEM NOTIFICATION -- "
config.template.css = "./style.css"
config.template.body = "stuff"
```
Also maybe `${HOME}/.newsletter-cli/config.js` instead of `${HOME}/.newsletter-cli.js`? That would give a reasonable place to stick the custom `style.css` too...

Happy to make adjustments to my PR. Thanks for this project, it's much nicer than the hack I've been using.

### CODE REVIEW

**Comment by thehunmonkgroup** [comment]
I went ahead and moved `${HOME}/.newsletter-cli.js` to `${HOME}/.newsletter-cli/config.js`, just makes sense.

Also replaced `custom_` with `template_` for the config variables, and upon further reflection, I think a single level of hierarchy is better for the custom config file anyways.

**Comment by thehunmonkgroup** [comment]
One more optimization and one more config option:
* Abstracted the user config loading to its own module, so it's now easily included in whatever other module needs access to it. Note that the linter was complaining about using require conditionally, but the code runs fine, so for now I disabled that rule on that line of code -- if there's some better way to do it I can adjust
* Added a `source_default` config option, which allows providing a default source email in the config file.

Pretty sure I'm done with the enhancements I need, I can also take a crack at tests and doc if that would help the PR along, but wanted to wait for feedback on what I have so far.

File: src/lib/UserConfig.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import fs = require("fs");
+import os = require("os");
+
+const homedir = os.homedir();
+const configModule = `${homedir}/.newsletter-cli/config`;
+
+export let userConfig = {} as any;
+if (fs.existsSync(`${configModule}.js`)) {
```
Comment (OrKoN) [inline src/lib/UserConfig.ts:8]: could you create a class UserConfig here that would have proper types? The class would read & parse the actual config .js file. Also, please move reading the config file out of the global scope and inject the instance of UserConfig to the specific commands in index.ts.

**Comment by OrKoN** [comment]
Could you include some documentation about into README?

File: src/lib/UserConfig.ts (hunk 1 above)
Comment (thehunmonkgroup) [reply src/lib/UserConfig.ts]: I've refactored to add a `UserConfig` class, but I'm not sure how you're wanting the injection to work from `index.js` -- are you wanting the instantiated `userConfig` object passed down through all object instantiations to the class that will eventually use it?

In that case, for example, the customizations for the `prepare` command would need to be passed to the creation of the `PrepareCommand` class and the `Newsletter` class.

Comment (OrKoN) [reply src/lib/UserConfig.ts]: Yes, passing it via constructor will be appropriate.

File: src/lib/UserConfig.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import fs = require("fs");
+import os = require("os");
+
+const homedir = os.homedir();
+const configModule = `${homedir}/.newsletter-cli/config`;
+
+export class UserConfig {
+  public load() {
```
Comment (OrKoN) [inline src/lib/UserConfig.ts:8]: please define the expected type of the config data. Something like:

```
export class UserConfig {
  private _config: unknown;
  constructor(path: string) {
     this._config = require(string);
  }
  
  getTemplateSubject(): string {
   const templateSubject = _config.template_subject;
    if (typeof templateSubject !== 'string') {
       throw new Error(...);
    }
    return templateSubject;
  }
};
```

Then the path can be resolved in index.ts and provided down to the classes that need the config.

Comment (thehunmonkgroup) [reply src/lib/UserConfig.ts:8]: OK, that adjustment is pretty straightforward. Another question: is it fine to pass the entire instantiated `UserConfig` object down through the various constructors, and let the destination class call the methods on that object? E.g., the `Newsletter` class would call `getTemplateSubject()`, etc? I think that will be cleaner, and still provide the type checking you want for the config data.

Comment (OrKoN) [reply src/lib/UserConfig.ts:8]: yes, that is what I meant

Comment (thehunmonkgroup) [reply src/lib/UserConfig.ts]: OK, I think this is complete per your request.

**Comment by thehunmonkgroup** [comment]
Note that I added one more config option: `ses_send_configuration`

I found this necessary as the SES -> SNS messages for failed render that you suggest in the README require that the AWS SDK sends a message with the configuration set declared, or SES will not publish to the SNS topic.

File: src/lib/SesTransport.ts [hunk 3]
```diff
@@ -7,10 +7,12 @@ process.env.AWS_SDK_LOAD_CONFIG = "1";
 export class SesTransport {
   private ses: aws.SES;
   private templates: Handlebars.TemplateDelegate[];
+  private userConfig: any;
 
-  constructor() {
+  constructor(userConfig: any) {
```
Comment (OrKoN) [inline src/lib/SesTransport.ts:12]: please use UserConfig type instead of any here and in other places.

File: src/lib/UserConfig.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+export type Config = {
+  template_subject?: string
+  template_css?: string
+  template_body?: string
+  source_default?: string
+  ses_send_configuration?: object
+}
+
+export class UserConfig {
+  private _config: Config;
```
Comment (OrKoN) [inline src/lib/UserConfig.ts:16]: this should be `unknown` so that TS checks that the type checking is correct.

Comment (thehunmonkgroup) [reply src/lib/UserConfig.ts:16]: I don't think this is quite right. When I try using unknown there, I get compiler errors:

```
npm run compile 

> newsletter-cli@0.4.0 compile /home/hunmonk/git/newsletter-cli
> npm run clean && tsc -p .


> newsletter-cli@0.4.0 clean /home/hunmonk/git/newsletter-cli
> rm -rf dist

[:~/git/newsletter-cli] custom-config-file(+0/-0) ± npm run compile 

> newsletter-cli@0.4.0 compile /home/hunmonk/git/newsletter-cli
> npm run clean && tsc -p .


> newsletter-cli@0.4.0 clean /home/hunmonk/git/newsletter-cli
> rm -rf dist

src/lib/UserConfig.ts:17:29 - error TS2571: Object is of type 'unknown'.

17     const templateSubject = this._config.template_subject;
                               ~~~~~~~~~~~~

src/lib/UserConfig.ts:24:25 - error TS2571: Object is of type 'unknown'.

24     const templateCss = this._config.template_css;
                           ~~~~~~~~~~~~

src/lib/UserConfig.ts:31:26 - error TS2571: Object is of type 'unknown'.

31     const templateBody = this._config.template_body;
                            ~~~~~~~~~~~~

src/lib/UserConfig.ts:38:27 - error TS2571: Object is of type 'unknown'.

38     const sourceDefault = this._config.source_default;
                             ~~~~~~~~~~~~

src/lib/UserConfig.ts:45:34 - error TS2571: Object is of type 'unknown'.

45     const sesSendConfiguration = this._config.ses_send_configuration;
                                    ~~~~~~~~~~~~


Found 5 errors.
```
I *think* the right approach is to declare `unknown` for all the `type Config` attributes, which seems to compile and work ok, and my latest push includes that fix.

Comment (OrKoN) [reply src/lib/UserConfig.ts:16]: alright, thanks!

**Pull request merged**